### PR TITLE
Crash on unknown config

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: starcluster/cluster.py [hunk 1]
```diff
@@ -436,8 +436,7 @@ def __init__(self,
                  config_on_master=False,
                  dns_suffix=None,
                  node_instance_array=[],
-                 impaired_threshold_sec=120,
-                 **kwargs):
```
Comment (FinchPowers) [inline starcluster/cluster.py:440]: Removed kwargs. It was not used and allowed bogus config to pass w/o notice.

File: starcluster/cluster.py [hunk 2]
```diff
@@ -638,6 +637,10 @@ def update(self, kwargs):
         for key in kwargs.keys():
             if hasattr(self, key):
                 self.__dict__[key] = kwargs[key]
+            else:
+                msg = "Unknown configuration key: " + key
+                log.error(msg)
+                raise Exception("Unknown configuration key: " + key)
```
Comment (FinchPowers) [inline starcluster/cluster.py:643]: When loading config from an existing cluster, log/crash on unknown settings.

File: starcluster/config.py [hunk 3]
```diff
@@ -563,7 +563,7 @@ def _load_instance_types(self, store):
         if isinstance(instance_types, basestring):
             return
         itypes = []
-        cluster_section['node_instance_types'] = itypes
+        cluster_section['node_instance_type'] = itypes
```
Comment (FinchPowers) [inline starcluster/config.py:566]: This patch revealed that this was a bug.

File: starcluster/config.py [hunk 4]
```diff
@@ -807,6 +807,8 @@ def get_cluster_template(self, template_name, tag_name=None,
             if not ec2_conn:
                 ec2_conn = self.get_easy_ec2()
 
+            del kwargs['__name__']
+            del kwargs['extends']
```
Comment (FinchPowers) [inline starcluster/config.py:811]: Values that should not make it to Cluster.**init**.

File: starcluster/static.py [hunk 5]
```diff
@@ -293,7 +293,6 @@ def create_sc_config_dirs():
     'cluster_size': (int, True, None, None, None),
     'cluster_user': (str, False, 'sgeadmin', None, None),
     'cluster_shell': (str, False, 'bash', AVAILABLE_SHELLS.keys(), None),
-    'subnet_id': (str, False, None, None, None),
```
Comment (FinchPowers) [inline starcluster/static.py:296]: Also a bug. Replaced with subnet_ids.

**Comment by jraby** [comment]
image

**Pull request merged**